Review : More groundwork

lqd opened the pull request:
More groundwork on:
- enums. The enum test now compiles and works.
- c-enums
- updated binaryen to get br_table support and the latest fixes
- with br_table, match expressions on variants were implemented (MIR switch, no switchInt yet). An example showcasing those 3 features was added as well.
- first pass at tuple assignments
- a slightly simplified cmp.rs from std/core compiles (without derives right now, because they require more type support than just i32s, and panic, etc)
- with enums, a simple version of Option can be compiled, so there's an example of that.
- a plan for the calling convention (which I'll have to check with @eddyb), especially for functions returning structs: temporarily hacked with copies to make it work first for review, until a more efficient/correct method is implemented
- modified some examples to have paths that match what #[derive] expects, and added Copy/Clone derivation to the struct example
- with Option and some of the previous work, a simplified example of IntoIter and Iterator works
- added successful compile-pass tests, mostly extracted from the current examples
- added xfail tests to document two features needing to be done (eg switchInt)
- fixed a couple latent bugs, related to binaryen's Duff's device, and monomorphizations recorded "as seen" (which I'll also have to check with @eddyb) to not trans again
- brought back the entry_fn to handle non #[main] main fns for example, as to not require the use of attributes

Marking the PR for review as I'm not sure we would want it to land as is, but it's been a week's worth of work already and I didn't want to delay the review of the other parts any longer. (eg I think we'd want to have a cleaner/correct calling convention implementation to land)

brson commented:
r? @eholk do you mind looking at this? You've been better at scrutinizing the actual code in these PR's than I.

brson commented:
This is so amazing. r=me if eholk is happy.

lqd commented:
@brson even after @eholk reviews the other parts, I'd still like to work on cleaning up the calling convention before merging, but I didn't want to delay review of all the other parts any longer :)

eholk commented:
Sorry, I was on vacation for the past few days. It'll be a little bit before I can review this, but I'll try to do it by the end of the week if that's alright.

lqd commented:
@eholk no problem really (the 14th's a bank holiday anyway), by then, I will have added (probably tomorrow) another couple commits to this PR to 1) add CLI parsing that will ease testing (and makes developing easier in general) as we talked about (but not yet assert! support as it seems Binaryen traps with a C++ exception, which I'm not sure can interact well from our side, but I'll investigate more) and 2) a first cleanup of the "memcpy" (using 1-8 bytes instead of just 1, because I only recently realized linear memory was little endian and offsets were in bytes... so that'll fix some of the weird behavior I was seeing in more complex cases than the ones I've been testing with so far). I definitely want to get rid of this copy soon (which will be transparent anyway, and doesn't impact the rest of your review) but I didn't have the time to do it yet.

lqd commented:
Here goes, crossing my fingers I don't mess up the history as I always do.

lqd commented:
@eholk here you go, those are the modifications I mentioned last time (it's hard to detect/debug wasm memory errors until it can be exported to a js uint8array or maybe even just inspected through the interpreter). The commits also include some additional cleanup, and fixing compiling operators.rs isize on x64 (maybe we should have automated tests on multiple os/arch in this repo as well eventually) — which made me wonder, since wasm doesn't have isize yet IIRC, I thought the best way for now to handle them was to force them to i32s at compile time until we handle more types, what do you think ?

Until you can review this, I'll probably look at assert! and panic! a bit more. On osx, binaryen trapping apparently does what we thought assert! would need, while on windows it showed a MS failure UI and didn't display any information about the error, so we'll see. (maybe it behaves differently there in Debug vs Release who knows)

kripken commented:
> since wasm doesn't have isize yet IIRC

Yes, current wasm is "wasm32", where the pointer size is 32-bit. So you can basically treat this like a 32-bit target where isize is 32-bit. (In the future a wasm64 is planned, however, it's long-term and also will probably be rare in practice.)

lqd commented:
@kripken yeah that's what I figured as well, thank you for confirming 😄

eholk commented on rust-examples/cmp.rs:
```diff
@@ -0,0 +1,382 @@
+    wasm::print_i32((x.ge(&y)) as i32);
+}
+
+pub mod marker {
```
Are these traits all copied from libcore/libstd, or are these reimplemented from scratch?

lqd replied:
as mentioned in the PR it's a slightly simplified cmp.rs from std/core :)
I just removed the derives (which come with their own requirements, eg assert_receiver_is_clone etc, which we may not end up wanting to port, but mostly because the generated impls for some of those traits required u64s and such :) and also I commented most of the cmp.rs impl macros for most if not all of the types other than i32. Quite a modification, but apart from that it's the regular rust traits.
The iterator.rs example however has a reimplementation of IntoInter and Iterator (since the iter support pulls a lot of other stuff which don't work yet)

eholk commented on rust-examples/enum.rs:
```diff
@@ -0,0 +1,63 @@
+    pub trait Copy : Clone { }
+}
+
+pub mod clone {
```
Do cross-file imports work yet? If so, it'd be good to share the implementation of these traits. I realize in the long run these will probably be handled by libstd, but it may be good to avoid some duplication in the short term. Feel free to save this for a future PR, or ignore entirely :)

lqd replied:
No they don't work yet. We're gonna have to do something about it, and think about how it should be exposed to wasm. Eg one rust module == one wasm module, or something else. A problem being wasm modules are not rust modules, and rust pub isn't exactly a wasm export, both are also designed to interop with other modules, or be called from javascript. That comes with additional constraints on which parameter types can be supported, etc. 
Still lots of questions left to answer before dealing with that I think (maybe after I get back from my august vacation), even though it is definitely an important topic.

eholk replied:
Sounds good. I just opened #29 for discussion about this issue.

eholk commented on rust-examples/fibonacci.rs:
```diff
@@ -107,7 +107,6 @@ impl Fibonacci for i32 {
     }
 }
 
-#[main]
```
Nice!

eholk commented on src/bin/mir2wasm.rs:
```diff
@@ -1,5 +1,4 @@
-#![feature(rustc_private, custom_attribute, link_args)]
-#![allow(unused_attributes)]
```
I'm glad you were able to get rid of `allow(unused_attributes)`. It seems good for code hygiene.

lqd replied:
yeah I think it was just a leftover from the miri bin and AFAICT we're not going to use miri with those attributes anyway :)

eholk commented on src/bin/mir2wasm.rs:
```diff
@@ -11,39 +10,75 @@ extern crate rustc_driver;
+            let entry_fn = state.session.entry_fn.borrow();
+            let entry_fn = if let Some((node_id, _)) = *entry_fn { Some(node_id) } else { None };
+            trans::trans_crate(&state.tcx.unwrap(), state.mir_map.unwrap(), entry_fn, &options)
                 .unwrap(); // FIXME
```
What is the FIXME here for? Is it that there is no error handling? Maybe add a comment clarifying that.

lqd replied:
I couldn't tell really as it's not my code :) but I think @brson meant that as requiring better error handling throughout the whole compiler, and here in particular where there's an unwrapped Result, no Errs, etc ?

eholk commented on src/bin/mir2wasm.rs:
```diff
@@ -11,39 +10,75 @@ extern crate rustc_driver;
+    let wasm_compiler_args = ["--run", "-O", "-q", "-h", "--help"];
+    let rustc_args : Vec<String> = std::env::args().filter(|arg| !wasm_compiler_args.contains(&arg.as_ref())).collect();
+
+    // TODO: use a command line parsing library
```
I was just about to say this same thing, but then I saw the TODO, so you beat me to it!

lqd replied:
yeah I didn't want to add a crate to the compiler without talking to you guys about it first, and it literally took 10 minutes to do by hand for those 3 flags (but OMG are they useful). With --run you could even already add the run-pass to your test runner I think

eholk replied:
Nice!

I forget, is there a standard (de facto or otherwise) command line parsing library? If so, I have absolutely no objections to using it here.

I'll have to try and get the run-pass tests going.

lqd replied:
I thought getopts was a very common one, especially since it's in the nursery and all, but there are others (clap comes to mind). What do you think @brson ?
Next time we need another CLI flag, it would be nice to switch to the chosen library at the same time.

eholk commented on src/bin/mir2wasm.rs:
```diff
@@ -11,39 +10,75 @@ extern crate rustc_driver;
+                             Options: \n    \
+                             -h, --help    Display this message \n    \
+                             -O            Optimize the compiled wast module \n    \
+                             --run         Run the compiled module through the interpreter, without printing it \n    \
```
I read the explanation for `--run` as "run the compiled module but silence stdout." What about saying "instead of printing it?"

lqd replied:
sure I'll change it (and commit once, possibly squashing it if you guys prefer, with all the changes for this review) (and probably tomorrow as it's 4AM here ;)
It doesn't exaclty silence stdout wholesale, it silences the compiler's stdout but not the interpreter's

eholk commented on src/trans.rs:
```diff
@@ -72,6 +107,10 @@ impl<'v, 'tcx: 'v> Drop for BinaryenModuleCtxt<'v, 'tcx> {
 
+// The address in wasm linear memory where we store the stack pointer
+// TODO: investigate where should the preferred location be
+const STACK_POINTER_ADDRESS : i32 = 0;
```
I realize that on IRC I kind of advocated for sticking the stack pointer at 0 just because we can and I think that's cool. From a software reliability standpoint, this is probably not the best idea. For pure Rust code, we're probably safe from accidentally dereferencing a null pointer, but if we ever link in something from C we could end up with a case where something that might be expected to crash would instead overwrite the stack pointer. This will probably lead to a crash later, so maybe it's not a big deal, but I could also see this opening up security problems (although the WASM sandbox will probably protect us somewhat here).

The nice thing about this being a constant is that it's trivial to change. We should keep an eye out for any unsafe pointer casts in real-world code (or as close as we have, anyway) that mir2wasm compiles, since these are where the risks will come from.

lqd replied:
yeah true. I just put it in there to try if it worked and it did, but as Alon mentioned it should be, say, at 8 :)

eholk commented on src/trans.rs:
```diff
@@ -145,11 +179,11 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
         // Translate arg and ret tys to wasm
         let binaryen_args: Vec<_> = self.sig.inputs.iter().map(|t| rust_ty_to_binaryen(t)).collect();
-        let mut needs_ret_local = false;
+        let mut needs_ret_var = false;
```
`needs_ret_var` seems like a better name to me, so I like this change.

lqd replied:
plus it's the more correct wasm terminology (and vars called locals, without calling params/args locals as well, caused me to miss the fact that relooper helper was using the wrong local index :)

eholk commented on src/trans.rs:
```diff
@@ -160,30 +194,33 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
 
-        debug!("{} wasm locals found", locals.len());
+        let locals_count = binaryen_args.len() + vars.len();
+        debug!("{} wasm locals initially found - params: {}, vars: {} (incl. stack pointer helper ${}, relooper helper ${})",
```
This line looks long. Is it under 100 characters? I believe that's the preferred line length for Rust code.

lqd replied:
probably not, but yeah we'll rustfmt eventually.
With rust's rightward drift even on my 13-inch laptop I sometimes go to 130 without noticing, and probably more at work...

eholk commented on src/trans.rs:
```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
+                    let adt_layout = self.type_layout(adt_ty);
+                    let discr_val = match *adt_layout {
+                        Layout::General { discr, .. } => {
+                            let discr_size = discr.size().bytes() as u32;
```
What sizes are possible here? Are they all supported? An `assert!` would be good to make sure any assumptions are explicit.

lqd replied:
yeah. 
IIRC <32bits

eholk commented on src/trans.rs:
```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
+                                BinaryenLoad(self.module, discr_size, 0, 0, 0, BinaryenInt32(), ptr)
+                            }
+                        }
+                        Layout::CEnum { .. } => {
```
Just to make sure I understand this right, CEnum layout means the enum is just a single value, and therefore passed by value instead of as a pointer to the tagged union?

lqd replied:
yes,  and you decide the value (but IIRC it has to be <32bits as well, or rather I wasn't able to make a 64bits one compile, the enums example + test has a c enum in it btw)

eholk commented on src/trans.rs:
```diff
@@ -226,9 +265,42 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
+                    // NOTE: plan for the calling convention: i32/i64 f32/f64 are to be passed using the wasm stack and function parameters.
+                    //       For the other types, the manual stack in linear memory will be used, and pointers into this stack
+                    //       passed as i32s. A call to a function returning a struct will require preparing the output return value space on
+                    //       the caller function's frame, and the called function will write its return value there to avoid memcpys
```
This calling convention seems good to me. When returning a struct, what position does the output pointer go in? (I'm guessing I'll find this out once I read a bit further)

lqd replied:
unfortunately you won't, I was doing that as we speak. It's just a big todo atm. Can I suggest reading the whole thing first and then going through it line by line ? Esp because a lot of concepts are scattered throughout the file, and the TODOs and comments are where they are usually the most important "only" (ie the function epilogue's code is before the prologue, etc)

lqd replied:
but to answer your question in advance, I was thinking: we could pass the output pointer, but would you see an issue if we just write to the frame pointer ? (reminder: I'm very inexperienced in rust, rustc, wasm, _and_ compiler programming so this might very well be just a stupid idea :)

eholk replied:
Next time I'll do a quick pass over the code before I do the line-by-line pass so my questions make more sense :)

I don't think there's anything inherently wrong with just writing to the frame pointer, although that's still a little undefined. I'm assuming you mean something like, before the function call, the stack looks like:

```
... other stuff ...
-----
caller data (locals, etc)
-----
```

Then during the call it looks like

```
... other stuff ...
-----
caller data (locals, etc)
-----
callee data (locals, etc)
-----
```

And finally afterwards you have: 

```
... other stuff ...
-----
caller data (locals, etc)
-----
callee return value
-----
```

This works, but it can have some downsides. The callee has to move all of its own data out of the way before it returns (although really it can just reserve space for its return value before putting anything else on the stack). The caller will also probably have to move the value to the place where it wants it to be.

Alternatively, the caller can just pass in an output pointer saying "stick your return value here." That way when the function returns the data is already in the right place.

Hopefully that all makes sense. I opened #30 that we can use to discuss the design of this area some more.

Also, don't worry about being inexperienced. You're doing some fine compiler engineering work here!

lqd replied:
Close to what I meant, but I guess I could have been clearer (esp. wrt the calling convention notes and todos) so I'll clarify this in the issue itself, so we don't have to fish around for PR comments, code comments, and so on!

eholk commented on src/trans.rs:
```diff
@@ -250,20 +322,78 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
                                 if b_fnty == BinaryenNone() {
                                     // The result of the Rust call is put in MIR into a tmp local,
                                     // but the wasm function returns void (like the print externs)
                                     // TODO: what representation should the unit type have in wasm ?
```
I'd say ideally the unit type should have no representation, which I guess means it's represented as a 0 byte value. If we have to give it a value, maybe 0 or -1 as an i32.

lqd replied:
since then I've tried the code the compiler outputs in the browser, and the validation rules are stricter, so I also think it probably shouldn't have a representation. I managed to change the compiler to ignore its assignments already, but we'll probably have to go through everything, make fns return void, and so on

eholk commented on src/trans.rs:
```diff
@@ -250,20 +322,78 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
-                                let (lvalue_index, _) = self.trans_lval(lvalue);
+                                let dest = self.trans_lval(lvalue);
 
                                 if b_fnty == BinaryenNone() {
```
Just to make sure, `b_fnty` is the return type, right?

lqd replied:
yes, and cleaning up as I find things, I've since renamed it to `ret_ty` :)

eholk commented on src/trans.rs:
```diff
@@ -250,20 +322,78 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
+                                    debug!("emitting {:?} Call to fn {:?} + SetLocal({}) for unit type", call_kind, func, dest.index.0);
                                     binaryen_stmts.push(b_call);
 
                                     let unit_type = BinaryenConst(self.module, BinaryenLiteralInt32(-1));
```
Ah, I see you went with -1. Good choice.

For the 0-byte representation, I think in practice what that means is that you'd basically just skip all these lines. No would should really be able to observe a void value, so it doesn't really matter what they are, as far as I know.

lqd replied:
exactly, like I just added in a previous comment about this, we're on same `void` page here :)

lqd replied:
and by "exactly" I actually mean not skipping all those lines but just the set_local, we'll still need to call the fn ;)

eholk commented on src/trans.rs:
```diff
@@ -446,8 +638,8 @@ impl<'v, 'tcx: 'v> BinaryenFnCtxt<'v, 'tcx> {
                 // TODO: works for refs to "our stack", but not the locals on the wasm stack yet
```
I'm interpreting this TODO as meaning we don't have the analysis that spills any address-taken variables onto onto the Rust stack. Is that correct?

lqd replied:
yes. so right now a mut i32 ref is in fine a copy...

eholk commented on src/trans.rs:
```diff
@@ -844,13 +1195,75 @@ fn rust_ty_to_binaryen<'tcx>(t: Ty<'tcx>) -> BinaryenType {
     s.chars().map(|c| {
         match c {
-            '<' | '>' | ' ' => '_',
+            '<' | '>' | ' ' | '(' | ')' => '_',
```
This can be in a followup PR, but we probably want this to make different characters to different strings. As it stands now, something like `foo<bar>` and `foo(bar)` both map to `foo_bar_`, which is also a name someone can type. This could potentially lead to weird bugs caused by name collisions.

I suspect using a reasonably common C++ name mangling scheme might be the right approach here.

lqd replied:
yeah, very good point

eholk commented on tests/compile-pass/enum.rs:
```diff
@@ -1,5 +1,3 @@
-// xfail
```
Hurray!

eholk commented:
Sorry about taking so long to finish looking over this. I'm really impressed with how many features this adds, and how quickly you've been able to implement so me!

r=me / lgtm / +1 / whatever the right way for us to indicate approval :)

brson commented:
Thanks @eholk ! \o/

lqd commented:
cool, thank you @eholk. I'll make another PR addressing the comments in this review soon.
This'll be without a doubt my last PR for a while as I'll be leaving on vacation soon.

By the time I get back, the work upgrading LLVM in Rust and Emscripten to get wasm support "for free" will be finished so that's exciting :smile:
